test.sh

```diff
@@ -128,57 +140,87 @@ headers_line() {
 }
 
 test404() {
-  curl -s -D "$SHUNIT_TMPDIR/headers" -o "$SHUNIT_TMPDIR/out" "http://$FICAI_LISTEN/derp"
-  assertEquals 'HTTP/1.1 404 Not Found' "$( headers_line 1 )"
+  request "http://$FICAI_LISTEN/derp"
+  assertStatus 'HTTP/1.1 404 Not Found'
+  assertError 'not_found' 'not found'
+}
+
+test405() {
+  request "http://$FICAI_LISTEN/v1/signals" -X PUT
+  assertStatus 'HTTP/1.1 405 Method Not Allowed'
+  assertError 'method_not_allowed' 'method not allowed'
```

Comment:
Why do we want a JSON error for this?

Reply:
Uniformity of handling, mostly. I don't expect any client to _intentionally_ handle a json 405. Helped when figuring out how warp's rejection system worked.

Reply:
Uniformity of handling where? So far the only affected place seems to be the unit tests, which doesn't sound like a great justification for more code. (That said, it's not too much more code, so I guess it's ok...)

Reply:
In the client code; of which the testsuite is the only example present in this repo. For one thing, it saves me when I'm running `curl ... | jq` against it and forget to update an `-XPATCH` at the beginning of the line when I change the URL, or fat finger the query string, etc. ;)

It seems strange to me to have an api that only returns json some of the time, and for some endpoints or error modes might just print arbitrary text. Properly handling content type negotiation seems like a much larger can of worms than just always returning json, personally.

Reply:
For those errors which are supposed to be handled by clients, sure, a uniform error format is great. But errors like "method not allowed" or "malformed request body" are programmer errors, and IMO effort should be spent on fixing them, not on making them look good.
Of course, it'd be a different story if there was some centralized framework code that made all errors look the same.

Reply:
I had no idea this would be such a contentious change; it's throwing me for a loop. Honestly thought it was an oversight and not an intentional design choice. Do you mind explaining your reasoning and argument a bit more so I can understand? 

You've said that it's "more code", are you worried about the amount of code in the server? Is this just be a fundamental disagreement where you think amount of code is more important and I think ergonomics and server-client contract complexity is more important? You said the couple of `if` branches here you'd be willing to accept, are you worried about code growth in the future?

> IMO effort should be spent on fixing them, not on making them look good

Do you think returning json here makes it _harder_ or _slower_ for a programmer on the other end to fix cases where they run into these errors? Do you think there's a place in the server where we could be spending effort to prevent these entirely, or some mutually exclusive way to make them easier/faster to deal with?

This ties into the error `code` discussion. It seems like the server-client contract you're shooting for is:
* when called correctly, return a well formed json response
* when called incorrectly, return an http status code and a raw string body suitable for a human

like (details omitted for brevity)
```
struct Signals;
enum Response {
    Success(Signals), // json
    Error(String), // NOT json
}
```
is that a correct interpretation?

Whereas my original intent was:
* return a well formed json response
* when called incorrectly, return an http status code, a specific error code, and a message suitable for a human

like
```
struct Signals;
enum Error {
    BadBody { message: String },
    BadQuery { message: String },
}
enum Response {
    Success(Signals), // json
    Error(Error), // json
}
```

I value strongly/richly typed interfaces and the adherence to principle of least surprise. Are these things you also value, just weigh differently compared to server code size or some other factor you haven't stated?

Reply:
> I had no idea this would be such a contentious change; it's throwing me for a loop.

I mean, I did say earlier: "That said, <...> I guess it's ok...", and even approved the PR. Now I'm just trying to explain my position, but I'm not going to prevent you from merging this code as is :)

----

That said, let's continue :D

> You've said that it's "more code", are you worried about the amount of code in the server?

Not quite. (For the following, I'd like us both to keep in mind that the project is very small right now, so in the absolute sense there is almost no difference between these approaches. I'm extrapolating the thought patterns I see behind the different approaches and thinking about how they could affect us if/when the project grows.) "More code" means "more effort spent writing it", and as a volunteer-only OSS project, we have a very tight effort budget :) So, when I started writing this server, I approached it with the mindset of "write the least amount of code and make the least amount of features required to make it all work (and produce the least amount of maintenance burden)". I am the main contributor (actually, the only one, for now) to another volunteer-only project, and literally the worst problem we have that's preventing us from doing a public release is the lack of manpower.

So, when I needed to signal an error, I did it in the easiest way I could imagine that got the job done: set a 4xx/5xx HTTP status (easy to see it's not a 2xx on both) and some text that would tell me what went wrong. I thought about how I'd handle these errors (both during development and in a client) and judged that making the error response body more complex, *on the scale of the project at the time*, wouldn't make it any easier for me to handle the errors.

Now we're introducing a new error format. To use it on the server, we now need to *remember* to write, and then to *actually* write, at least one more line in each API method — either `json_or_error` or `.header(CONTENT_TYPE, "application/json")`. The latter case also needs a correct body — which this PR neglects to do, by the way. (I totally understand why! I'm not saying this as a slight to you!) The *former* case needs a bunch of additional lines for each error type — and, again, some errors are covered (like the method-not-allowed one; I'm guessing you ran into it while testing), but others aren't, there's a whole bunch of other HTTP codes that can probably be generated by our under some obscure conditions. As I said earlier:

> Of course, it'd be a different story if there was some centralized framework code that made all errors look the same.

----

Now, to address the handling of errors. On the client, *unless we bury the API calling code beneath a framework layer* (handmade or a third-party library), the error handling would look like either of these (pseudocode similar to what I remember of the actual JS code I had to write all those months ago):
```
if (response.code != 200) {
  doContextSpecificHandlingButProbablyJustShowAnAlert(response.text);
}

if (response.code != 200) {
  doContextSpecificHandlingButProbablyJustShowAnAlert(response.json.message);
}
```
At least, this is what it looked like when I last took a look at the client-side code :) The differences are insignificant. But, of course, if we *do* now use a framework that can take advantage of all errors being JSON-formatted, then, sure, we should convert them all.

> Do you think returning json here makes it harder or slower for a programmer on the other end to fix cases where they run into these errors?

No, it's basically the same IMO.

> Do you think there's a place in the server where we could be spending effort to prevent these entirely, or some mutually exclusive way to make them easier/faster to deal with?

Well, I mentioned a few times now that I'd be happy with some kind of "single block of code" on the server that would handle all current and future errors and present them in a "blessed" way, and that would make it as easy as possible to write new APIs with support for it. But I suspect you misunderstood me when I said "effort should be spent on fixing them", "them" being "programmer errors". Obviously encountering them is inevitable, like e. g. mistyping a URL and getting a 404. But when I get one, *regardless of what it looks like*, I just say "oh, duh!" and fix the URL and that's it.

----

> I value strongly/richly typed interfaces and the adherence to principle of least surprise. Are these things you also value, just weigh differently compared to server code size or some other factor you haven't stated?

I value strong typing (and, really, any other thing) where (I think) it helps me solve or avoid some kind of a problem in a situation which I can't sidestep altogether. With APIs, the principle of least surprise ends for me when good documentation appears (which I saw in the other PR, and I approve it wholeheartedly). There's nothing to be surprised about if the documentation warns you about all the formats beforehand :) Now, of course, my annoyance would grow rather quickly if I had to always write new complex parsing code to handle the errors; then I'd want to be able to make a "parseError" method and call it everywhere, and a common format would help me with that.

----

With all that said; I recognize that all my arguments here are rather weak and a matter of personal preference. So, since you're the one actively contributing code right now, you get to decide how we write it going further :)

Reply:
Gotcha, I think I generally get where you're coming from and what you're angling towards now; I appreciate you expounding :)

I read the "I guess it's ok..." as very reluctant, particularly since it was sent _after_ approval, but probably read too much into it because you were trying to address future concerns and I was worried if you were objecting to the bit of code that is already here that there were going to be many more objections in the future. I'm sure there's some catchy phrases for "two parties reacting to what they think the other party is going to do and entering a loop" ;) Even if you make the comments after you approve, I'll try to read less into the comments.

> To use it on the server, we now need to _remember_ to write, and then to _actually_ write...
> ...I'd be happy with some kind of "single block of code"...

I think this can actually be simplified further; I plan to give it another shot after the current queue of changes are fully merged at it's a little easier to see the patterns. Ran into some trouble with the orphan rule and private warp definitions, and decided some copy+paste in the mean time is easier than spending more time or jumping to proc macros or the like.

> ...or `.header(CONTENT_TYPE, "application/json")`. The latter case also needs a correct body — which this PR neglects to do, by the way.

I've gone blind by looking at my own changes too long, where are you referring to? I'm particularly worried that the `test.sh` line which checks to see if the body is valid json is written wrong or obvious test coverage is missing.

> There's nothing to be surprised about if the documentation warns you about all the formats beforehand :)

Even if the stakes here are much lower, coming from the c++ world, I know simply documenting sharp corners are not actually enough to prevent people from cutting themselves :p I think there's something to be said for simpler to communicate contracts as well -- where not having to qualify all endpoints as "might return arbitrary text" is my preferred outcome.

> With all that said; I recognize that all my arguments here are rather weak and a matter of personal preference. So, since you're the one actively contributing code right now, you get to decide how we write it going further :)

Please keep chiming in! It's opinions all the way down, and we're trying to achieve the same things from pretty close angles.

Reply:
> I've gone blind by looking at my own changes too long, where are you referring to?

... I went to check and realized that I was wrong there — for some reason I thought that the empty JSON objects that were added in `usermgmt.rs` were for errors. But no, those are for the good cases. Sorry, my bad :)